quickwit-actors/src/actor_handle.rs

```diff
@@ -0,0 +1,165 @@
+use crate::actor_state::ActorState;
+use crate::mailbox::Command;
+use crate::{Actor, ActorContext, ActorTermination, Observation};
+use std::fmt;
+/// An Actor Handle serves as an address to communicate with an actor.
+///
+/// It is lightweight to clone it.
+/// If all actor handles are dropped, the actor does not die right away.
+/// It will process all of the message in its mailbox before being terminated.
+///
+/// Because `ActorHandle`'s generic types are Message and Observable, as opposed
+/// to the actor type, `ActorHandle` are interchangeable.
+/// It makes it possible to plug different implementations, have actor proxy etc.use std::fmt;
+use tokio::sync::{oneshot, watch};
+use tokio::task::{JoinError, JoinHandle};
+use tokio::time::timeout;
+use tracing::error;
+
+pub struct ActorHandle<A: Actor> {
+    actor_context: ActorContext<A>,
+    last_state: watch::Receiver<A::ObservableState>,
+    join_handle: JoinHandle<ActorTermination>,
+}
+
+impl<A: Actor> fmt::Debug for ActorHandle<A> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("ActorHandle")
+            .field("name", &self.actor_context.actor_instance_name())
+            .finish()
+    }
+}
+
+impl<A: Actor> ActorHandle<A> {
+    pub(crate) fn new(
+        last_state: watch::Receiver<A::ObservableState>,
+        join_handle: JoinHandle<ActorTermination>,
+        ctx: ActorContext<A>,
+    ) -> Self {
+        let mut interval = tokio::time::interval(crate::HEARTBEAT);
+        let actor_instance_name = ctx.actor_instance_name().to_string();
+        let ctx_clone = ctx.clone();
+        tokio::task::spawn(async move {
+            interval.tick().await;
+            while ctx.kill_switch().is_alive() {
+                interval.tick().await;
+                if !ctx.progress().harvest_changes() {
+                    if ctx.get_state() == ActorState::Terminated {
+                        return;
+                    }
+                    error!(actor=%actor_instance_name, "actor-timeout");
+                    ctx.kill_switch().kill();
+                    return;
+                }
+            }
+        });
+        ActorHandle {
+            join_handle,
+            last_state,
+            actor_context: ctx_clone,
+        }
+    }
+
+    /// Process all of the pending message, and returns a snapshot of
+    /// the observable state of the actor after this.
+    ///
+    /// This method is mostly useful for tests.
+    ///
+    /// To actually observe the state of an actor for ops purpose,
+    /// prefer using the `.observe()` method.
+    ///
+    /// This method timeout if reaching the end of the message takes more than an HEARTBEAT.
+    /// Hence, it is only useful or unit tests.
+    pub async fn process_pending_and_observe(&self) -> Observation<A::ObservableState> {
```

Comment:
since it is unit tests only, it could make sense to build a helper function using public methods, for a more realistic scenario

Reply:
I did not understand your comment.

This method can be useful outside of unit tests.

Reply:
Then the comment is incorrect
`/// Hence, it is only useful or unit tests.`

Reply:
Oh yes! there are two lines saying different things. Thanks.